### PR TITLE
Add bisection method for `estimate_depletion_voltage`

### PR DESCRIPTION
**Pull request by yxzh321**
The bisection method is implemented in function `estimate_depletion_voltage`.
The previous `estimate_depletion_voltage` is called `old_estimate_depletion_voltage` now.

### CODE REVIEW

**Comment by fhagemann** [comment]
It looks like all of the tests that include determining the depletion voltage are failing.
@yxzh321 Can you look into this and see how far the new results are off from the old ones?

File: test/runtests.jl [hunk 1]
```diff
@@ -43,7 +43,7 @@ T = Float32
         @test isapprox( signalsum, T(2), atol = 5e-3 )
         nt = NamedTuple(sim)
         @test sim == Simulation(nt)
-        deplV = estimate_depletion_voltage(sim, (verbose = false,))
+        deplV = estimate_depletion_voltage(sim, verbose = false)
         @test isapprox(deplV, 1870*u"V", atol = 5.0*u"V")
```
Comment (fhagemann) [inline test/runtests.jl:47]: `:InvertedCoax` depletion voltage is around 1870V.

File: test/runtests.jl [hunk 2]
```diff
@@ -131,7 +131,7 @@ T = Float32
         signalsum *= inv(ustrip(SolidStateDetectors._convert_internal_energy_to_external_charge(sim.detector.semiconductor.material)))
         @info signalsum
         @test isapprox( signalsum, T(2), atol = 5e-3 )
-        @test isapprox(estimate_depletion_voltage(sim, (verbose = false,)), T(-13.15)*u"V", atol = 1.0u"V")
```
Comment (fhagemann) [inline test/runtests.jl:134]: `:Hexagon` depletion voltage is around -13.15V.

File: test/runtests.jl [hunk 3]
```diff
@@ -117,11 +117,11 @@ T = Float32
             l = vcat(wp.data[rmin:-1:2,φidx2,zidx], wp.data[1:rmax,φidx1,zidx])
             @test all(diff(l) .< 0) # the weighting potential should monotonously decrease from rmin to rmax
         end
-        @test isapprox(estimate_depletion_voltage(sim, (verbose = false,)), 0u"V", atol = 0.2u"V") # This detector has no impurity profile
```
Comment (fhagemann) [inline test/runtests.jl:120]: `:BEGe` has no impurities, which is why the depletion voltage should be 0V.

**Review by fhagemann — COMMENTED** [review]
The values used in the tests as comparison values are determined using the brute-force and computationally expensive method where for each voltage, the depletion status of the detector is calculated from zero and the depletion voltage is determined using a bisection algorithm.

For future reference:

```julia
cid = SolidStateDetectors.determine_bias_voltage_contact_id(sim.detector)
verbose = true
Umax = 3500.0
Umin = 0.0
first = true
done = false

@assert Umax * Umin >= 0

if Umin * Umax == 0
    sim.detector = SolidStateDetector(sim.detector, contact_id = cid, contact_potential = 0)
    calculate_electric_potential!(sim, device_array_type = CuArray, depletion_handling = true,
        refinement_limits = [0.2,0.1,0.05,0.03,0.02], verbose = false)
    if is_depleted(sim.point_types)
        done = true
        Umid = zero(T) * u"V"
    end
end

if !done
    while first || !is_depleted(sim.point_types)
        if !first
            Umax *= 2
            verbose && @info "Increasing Umax to $(Umax)"
        end
        first = false
        sim.detector = SolidStateDetector(sim.detector, contact_id = cid, contact_potential = Umax)
        calculate_electric_potential!(sim, device_array_type = CuArray, depletion_handling = true,
            refinement_limits = [0.2,0.1,0.05,0.03,0.02], verbose = false)
    end

    while abs(Umin - Umax) > 0.1

        Umid = (Umin + Umax)/2
        verbose && println("Searching in the range from $(Umin) to $(Umax)")

        sim.detector = SolidStateDetector(sim.detector, contact_id = cid, contact_potential = Umid);
        calculate_electric_potential!(sim, device_array_type = CuArray, depletion_handling = true,
                refinement_limits = [0.2,0.1,0.05,0.03,0.02], verbose = false)
           # ssd_write("unseg_$(U)V_linear_r_z_with_alpha_$(value[2]).h5", sim)

        if is_depleted(sim.point_types)
            Umax = Umid
        else
            Umin = Umid
        end
    end

    Umid = (Umin + Umax)/2
end

Umid
```

**Review by fhagemann — APPROVED** [review]

**Comment by fhagemann** [comment]
Checking the example in issue #324 now gives something more reasonable.

```julia
using SolidStateDetectors
T = Float32
sim = Simulation{T}(SSD_examples[:BEGe])
sim.detector = SolidStateDetector(sim.detector, SolidStateDetectors.ConstantImpurityDensity{T}(1e16)) # 1e10 cm^-3
calculate_electric_potential!(sim, depletion_handling = true, refinement_limits = [0.2,0.1,0.05,0.03,0.01])
estimate_depletion_voltage(sim)
```

now gives `+2347.847f0 V` instead of `-537625.5f0 V` which is way more reasonable.

**Comment by oschulz** [comment]
We could store the reference value in a text/JSON file and keep the code above in a script "update_depletion_reference_values.jl" (or similar) that rewrites that JSON file if run explicitly.

**Comment by fhagemann** [comment]
We could do that but I don't see why these values should ever change in the future (?)

**Pull request merged**